source/store/AppCompatibilityContext.tsx

```diff
@@ -0,0 +1,99 @@
+import getApplicationVersionStatus from "app/services/ApplicationVersionService";
+import {
+  AppCompatibilityContextType,
+  AppCompatibilityVisitor,
+  ApplicationCompatibilityState,
+  APPLICATION_COMPATIBILITY_STATUS,
+} from "app/types/AppCompatibilityTypes";
+import VERSION_STATUS from "app/types/VersionStatusTypes";
+import React, { useContext, useEffect, useState } from "react";
+import AppContext from "./AppContext";
+
+interface AppCompatibilityGuardProps {
+  children: React.ReactNode | React.ReactNode[];
+}
+
+// Create value of context with status and a full
+// visitor implementation
+const createAppCompatibilityContextValue = (
+  state: ApplicationCompatibilityState
+) => ({
+  visit<T>(visitor: Partial<AppCompatibilityVisitor<T>>): T | undefined {
+    switch (state.status) {
+      case APPLICATION_COMPATIBILITY_STATUS.COMPATIBLE:
+        return visitor?.compatible?.();
+      case APPLICATION_COMPATIBILITY_STATUS.PENDING:
+        return visitor?.pending?.();
+      case APPLICATION_COMPATIBILITY_STATUS.INCOMPATIBLE:
+        return visitor?.incompatible?.(state);
+      default:
+        return undefined;
+    }
+  },
+});
+
+// Hook for requesting application version status
+const useCompatibilityHook = () => {
+  const { isDevMode } = useContext(AppContext);
+  const [state, setState] = useState<ApplicationCompatibilityState>({
+    status: APPLICATION_COMPATIBILITY_STATUS.PENDING,
+    updateUrl: "",
+  });
+  const [versionPromise, setVersionPromise] = useState<Promise<any> | null>(
+    null
+  );
+
+  useEffect(() => {
+    // we have at most one outstanding call to fetching application version status
+    if (versionPromise) {
+      return;
+    }
+    setVersionPromise(async () => {
+      if (isDevMode) {
+        // in dev mode we defer the whole update logic to mimic
+        // production behaviour
+        return setState({
+          updateUrl: "https://www.example.com/placeholder",
+          status: APPLICATION_COMPATIBILITY_STATUS.COMPATIBLE,
+        });
+      }
+
+      const { status, updateUrl } = await getApplicationVersionStatus();
+      const url = String(updateUrl || "");
+      switch (status) {
```

Comment:
This Switch case feel a little bit overkill. And also a little bit dry.

wouldn't this be enough:

```js
  const compabilityStatus =
    status === VERSION_STATUS.UPDATE_REQUIRED
      ? APPLICATION_COMPATIBILITY_STATUS.INCOMPATIBLE
      : APPLICATION_COMPATIBILITY_STATUS.COMPATIBLE;

  return setState({
    updateUrl: url,
    status: compabilityStatus,
  }); 
```